File: matlab/fermifab/tensor_op.m
function y = tensor_op(x,N)
%TENSOR_OP - Operator tensor product
%
%   y = TENSOR_OP(x,N)
%
%   Calculate the tensor product operator
%   (x \otimes x ... \otimes x): wedge^N H -> wedge^N H
%   for a 1-body operator x: H -> H
%
%   Copyright (c) 2008-2015, Christian B. Mendl
%   All rights reserved.

assert(isa(x,'fermiop'));
assert(sum(get(x,'pFrom'))==1 && sum(get(x,'pTo'))==1,'Input must be a single-particle operator.');

y = fermiop(get(x,'orbs'),N,N,calc_tensor_op(full(get(x,'data')),N),get(x,'orbnames'));

File: python/fermifab/tensor_op.py
from scipy.sparse import csr_matrix
from .fermiop import FermiOp
import fermifab.kernel

__all__ = ['tensor_op']


def tensor_op(op, N):
    """
    Calculate the matrix representation of the N-fold tensor product of an operator.

    Args:
        op: quantum operator of type `FermiOp`, with `pFrom` and `pTo` equal to 1
        N:  number of tensor factors

    Returns:
        FermiOp: N-fold tensor product
    """
    if op.pFrom != 1 or op.pTo != 1:
        raise ValueError('operator particle numbers must be equal to 1')
    dims, val, ind = fermifab.kernel.tensor_op(op.data, N)
    # finally convert to dense matrix, for simplicity
    AN = csr_matrix((val, (ind[:, 0], ind[:, 1])), shape=dims).todense()
    return FermiOp(op.orbs, N, N, data=AN)
